### ISSUE TITLE
[Question] How to configure an app to access a port in localhost?

### ISSUE THREAD

**Issue by cakeslice** [caprover/caprover#1361]
I'm trying to setup Datadog for one of my CapRover apps and I need to access localhost:8126 (on the host itself) from within the container.

I looked online and some suggestions include using "host.docker.internal" as the hostname in the container and adding "--add-host=host.docker.internal:host-gateway" to the docker run command.

Is this possible with CapRover and if yes, how so?

**Comment by maietta** [comment]
Your request is a bit more advanced but definitely do-able. Being that Caprover does offer a service override textarea field, we may be able to leverage that. Firstly though, let's see what related issues are already present in this repository.

Related issue? https://github.com/caprover/caprover/issues/1325
Also related? https://github.com/caprover/caprover/issues/1005

Also, I referenced this document to find the taxonomy and nesting of the "service create" stuff from Docker's official docs, which can seem overwhelming to most people. https://docs.docker.com/engine/api/v1.40/#operation/ServiceCreate

Also I referenced https://caprover.com/docs/service-update-override.html.

To get a list of networks on the host, I SSH'ed into my Caprover testing VPS and ran `docker network ls`, which in my case looked like this:

root@vultr:~# docker network ls
NETWORK ID     NAME                      DRIVER    SCOPE
0db6cca69cd0   bridge                    bridge    local
mkhih3o8yw5d   captain-overlay-network   overlay   swarm
79b6c547465f   docker_gwbridge           bridge    local
95bdb20d337e   host                      host      local
bn0ssc9twjp3   ingress                   overlay   swarm
952e5e273dcc   none                      null      local

Since that is a stock Caprover server install, my guess is that "host" is the correct network, but I don't know for absolute certain.

```
TaskTemplate:
    Networks:
      Target: "host"
```

That placed into your App's service override section *might* work, but I am not well versed in this. Hopefully we're getting you close to the goal post.

**Comment by cakeslice** [comment]
Hello @maietta, thanks for the suggestion.

It seems to me that is just changing the container to **run on the host** which is not my objective.

What I need to do is to keep everything as normal, except map a port from **localhost:8126** to **container:8126** or have the host directly available like **host.docker.internal:8126**

Maybe it's possible do such mapping in the **captain-overlay-network** itself?

**Comment by maietta** [comment]
Okay, not sure how the wires were crossed but we're not going to get closer to the goal post without either a bunch of random trial and error or rephrasing the problem to better understand the goal.

It now sounds to me like you need to bind a port from the container to the host, which is dead simple in Caprover.

If that's the case, just use the App Config section of your app to add a Port mapping from your host, to point to your container. 
Once that is done, you should be able to reach your container from srv-captain--appname:8126 or https://appname.apps.yourdomain.com:8126. 

Here is what my "mail" app service port mapping looks like to bind various ports to my service containers.

![image](https://user-images.githubusercontent.com/11745020/156634933-e1c21eff-ee4b-40e6-aeb1-9eb872d43609.png)

**Comment by cakeslice** [comment]
It's actually the opposite, I don't need to access a container, I need to access the host from the container

The Port Mapping is mapping a port from the container to the host.

I don't need this feature urgently so I will try to find a solution later. If I'm able to solve I'll update this issue.

**Comment by githubsaturn** [comment]
@cakeslice - CapRover doesn't use docker containers, it uses Docker services (which are essentially orchestrated contianers). So some flags are a bit different. In this case, I think what you are looking for is --host-add

```
docker service update --host-add foo:127.0.0.2  srv-captain--yourappname
```

You can view all docker update flags here: https://docs.docker.com/engine/reference/commandline/service_update/

**Comment by mitkodkn** [comment]
@cakeslice Did you manage to get this to work?

**Comment by srijan** [comment]
For anyone in the future, I was able to do this using service update override:

```
TaskTemplate:
  ContainerSpec:
    Hosts:
      - "host-gateway host.docker.internal"
```

and by allowing the docker_gwbridge IP range in the firewall.

**Comment by niocncn** [comment]
@srijan Could you describe in more detail what you did in order to get access from the container to the ports on the host?

**Comment by SopanhaKok** [comment]
I was struggling with this problem as well. I did manage to access from the container by doing these:
1. using service update override as @srijan  suggested.
2. `docker network inspect captain-overlay-network` and get the ip subnet.
3. allow the ip subnet in the firewall

**Comment by pcace** [comment]
> 3. allow the ip subnet in the firewall
i know its a little bit offtopic, but in what scenario would i need to add what to the firewall? if i have only a ufw running protecting me from access from outside is there a need to change anything? 

thanks a lot for that solution!

**Comment by SopanhaKok** [comment]
If you need to allow external access to services running in containers (e.g., a web server or database accessible from outside the Docker host), you must configure the host firewall to allow incoming traffic on the respective ports.

and yes `sudo ufw allow from 10.0.1.0/24` should do the job

**Comment by Turab** [comment]
> @cakeslice - CapRover doesn't use docker containers, it uses Docker services (which are essentially orchestrated contianers). So some flags are a bit different. In this case, I think what you are looking for is --host-add
> 
> ```
> docker service update --host-add foo:127.0.0.2  srv-captain--yourappname
> ```
> 
> You can view all docker update flags here: https://docs.docker.com/engine/reference/commandline/service_update/

Anyone needs a graphical interface to manage hosts of the "service" can use Portainer. Go to services, find the service corresponding your container (app), and manage hosts entries per your wish. Since this is a service configuration, adding hosts here should prevail after container restarts.

**Comment by pcace** [comment]
Hi there, i am strugeling to achieve the same thing: 
i have a service running on the local machine on : http://localhost:8080/asdf
now i want to have a caprover container reverse proxy to be able to access the local service. 
so i did: 
`ufw allow from 10.0.1.0/24`
then
created a on click reverse proxy app, and added this service override: 
```
TaskTemplate:
  ContainerSpec:
    Hosts:
      - "host-gateway host.docker.internal"
```

and in the `UPSTREAM_HTTP_ADDRESS` env i use: `http://host.docker.internal:8080`

when i access the caprover nginx service i get: 

2025-05-06T15:28:19.403181052Z 2025/05/06 15:28:19 [error] 12#12: *1 host.docker.internal could not be resolved (3: Host not found), client: 10.0.1.5, server: _, request: "GET /asdf/ HTTP/1.0", host: "asdf.domain.com"
2025-05-06T15:28:19.403398917Z 10.0.1.5 - - [06/May/2025:15:28:19 +0000] "GET /asdf/ HTTP/1.0" 502 157 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:138.0) Gecko/20100101 Firefox/138.0" "89.247.166.106"

what am i missing here? 

thanks a lot!

**Comment by Turab** [comment]
@pcace , Caprover uses Docker Swarm and in Docker Swarm, `docker.host.internal` is not supported. Some people say adding this to your hosts file does the trick: `172.17.0.1 docker.host.internal`. Or you could assign IP addresses to your containers manually or as I mentioned above, you can use Portainer to edit services to add hosts on a graphical interface.

If one of these doesn't work, you should further explain what you want to achieve.
